libsolidity/codegen/ContractCompiler.cpp

```diff
@@ -343,6 +362,12 @@ void ContractCompiler::appendFunctionSelector(ContractDefinition const& _contrac
 	}
 
 	FunctionDefinition const* fallback = _contract.fallbackFunction();
+	solAssert(!_contract.isLibrary() || !fallback, "Libraries can't have fallback functions");
+
+	bool hasPayableFun = hasPayableFunctions(_contract);
+	if (!hasPayableFun && !interfaceFunctions.empty() && !_contract.isLibrary())
+		appendCallValueCheck();
```

Comment:
I think we should have a single bool that is true if and only if we already appended the call value check. If we change the condition here, we might forget changing the other conditions below.